File: src/dashboard/static/js/bots.js
import { api } from './api.js';
import { showToast } from './ui.js';

export async function createBot(e) {
  e.preventDefault();
  const accountId = document.getElementById('setupAccount').value;
  if(!accountId) return showToast('Vui lòng chọn hoặc tạo Tài khoản API trước', 'error');

  let params = {};
  try {
    params = JSON.parse(document.getElementById('setupParams').value);
  } catch(err) {
    return showToast('Lỗi JSON trong ô Tham số', 'error');
  }

  const symbolsInput = document.getElementById('setupSymbols').value;
  const symbols = symbolsInput.split(',').map(s => s.trim()).filter(s => s);

  const data = {
    name: document.getElementById('setupName').value,
    account_id: accountId,
    strategy_name: document.getElementById('setupStrategy').value,
    symbols: symbols,
    parameters: params
  };

  try {
    await api.createBot(data);
    showToast('Tạo Bot thành công!', 'success');
    window.showPage('mybots');
  } catch (err) {
    showToast('Lỗi tạo bot', 'error');
  }
}

export async function fetchBots() {
  const bots = await api.getBots();
  const grid = document.getElementById('botsGrid');
  
  if (bots.length === 0) {
    grid.innerHTML = '<div class="empty-state"><h3>Chưa có Bot nào đang chạy</h3><p style="margin-top:10px;">Vui lòng sang tab "Chiến lược giao dịch" để khởi tạo Bot đầu tiên!</p></div>';
    return;
  }
  
  grid.innerHTML = bots.map(bot => {
    const isRunning = bot.status === 'running';
    const statusClass = isRunning ? 'status-running' : 'status-stopped';
    const toggleBtn = isRunning 
        ? `<button class="btn btn-danger btn-sm" onclick="toggleBot(${bot.id}, 'stopped')">⏹ Dừng Bot</button>`
        : `<button class="btn btn-success btn-sm" onclick="toggleBot(${bot.id}, 'running')">▶ Chạy Bot</button>`;

    const symbolsStr = (bot.symbols || []).join(', ');

    // Job behavior settings — defaults true nếu undefined
    const allowEntry  = bot.allow_new_entry  !== false;
    const notifyEntry = bot.notify_entry     !== false;
    const allowExit   = bot.allow_exit_scan  !== false;
    const notifyExit  = bot.notify_exit      !== false;

    // Khi bot stopped, allow_new_entry luôn false và không cho toggle
    const entryDisabled = !isRunning ? 'disabled title="Bot đang dừng — không vào lệnh mới"' : '';

    function toggleHtml(label, value, field, disabled = '') {
      const checked = value ? 'checked' : '';
      const cls = value ? 'toggle-on' : 'toggle-off';
      return `
        <label class="bot-toggle ${cls}" ${disabled}>
          <input type="checkbox" ${checked} ${disabled}
            onchange="updateBotSetting(${bot.id}, '${field}', this.checked)"
            style="display:none">
          <span class="toggle-track"></span>
          <span class="toggle-label">${label}</span>
        </label>`;
    }

    return `
      <div class="card" id="bot-card-${bot.id}">
        <div class="card-header">
          <span class="bot-name">#${bot.id} ${bot.name}</span>
          <span class="bot-status ${statusClass}">${bot.status.toUpperCase()}</span>
        </div>
        <div class="stat-row"><span class="stat-label">Chiến thuật:</span><span>${bot.strategy_name}</span></div>
        <div class="stat-row"><span class="stat-label">Symbols:</span><span>${symbolsStr}</span></div>
        <div class="stat-row"><span class="stat-label">PnL:</span><span style="color: ${bot.total_pnl >= 0 ? '#0ecb81' : '#f6465d'}">${bot.total_pnl}</span></div>
        <div class="stat-row"><span class="stat-label">Win Rate:</span><span>${bot.win_rate}%</span></div>

        <!-- Job behavior settings -->
        <div style="margin-top: 12px; padding: 10px; background: rgba(0,0,0,0.15); border-radius: 6px; border: 1px solid var(--border);">
          <div style="font-size: 11px; color: var(--text-secondary); margin-bottom: 8px; text-transform: uppercase; letter-spacing: 0.5px;">⚙️ Cài đặt Job</div>
          <div style="display: grid; grid-template-columns: 1fr 1fr; gap: 6px;">
            ${toggleHtml('Vào lệnh mới', allowEntry, 'allow_new_entry', entryDisabled)}
            ${toggleHtml('Noti entry', notifyEntry, 'notify_entry')}
            ${toggleHtml('Quét đóng lệnh', allowExit, 'allow_exit_scan')}
            ${toggleHtml('Noti đóng lệnh', notifyExit, 'notify_exit')}
          </div>
        </div>

        <div style="margin-top: 15px; display: flex; gap: 10px;">
          ${toggleBtn}
          <button class="btn btn-sm" style="background: var(--border); color: #fff;" onclick="deleteBot(${bot.id})">🗑 Xóa</button>
        </div>
      </div>
    `;
  }).join('');
}

export async function toggleBot(id, status) {
  try {
    await api.updateBotStatus(id, status);
    showToast(`Bot ${status === 'running' ? 'đã khởi động' : 'đã dừng'}`, 'success');
    fetchBots();
  } catch (err) {
    showToast('Lỗi cập nhật trạng thái bot', 'error');
  }
}

export async function updateBotSetting(botId, field, value) {
  try {
    await api.updateBotSettings(botId, { [field]: value });
    showToast(`Đã cập nhật: ${_settingLabel(field)} = ${value ? 'Bật' : 'Tắt'}`, 'success');
    // Không reload toàn bộ grid — chỉ update visual state của toggle
    const card = document.getElementById(`bot-card-${botId}`);
    if (card) {
      const label = card.querySelector(`input[onchange*="'${field}'"]`)?.closest('.bot-toggle');
      if (label) {
        label.classList.toggle('toggle-on', value);
        label.classList.toggle('toggle-off', !value);
      }
    }
  } catch (err) {
    showToast('Lỗi cập nhật cài đặt', 'error');
    // Revert checkbox
    fetchBots();
  }
}

function _settingLabel(field) {
  const map = {
    allow_new_entry: 'Vào lệnh mới',
    notify_entry: 'Noti entry',
    allow_exit_scan: 'Quét đóng lệnh',
    notify_exit: 'Noti đóng lệnh',
  };
  return map[field] || field;
}

export async function deleteBot(id) {
  if(!confirm('Bạn có chắc muốn xóa bot này? Lịch sử PnL vẫn được giữ lại.')) return;
  await api.deleteBot(id);
  fetchBots();
}

File: src/dashboard/routers/bots.py
from fastapi import APIRouter, HTTPException
from sqlalchemy import select
from src.database.db import get_db
from src.database.models import Bot
from src.dashboard.schemas import BotCreate, BotStatusUpdate, BotSettingsUpdate

router = APIRouter(prefix="/api/bots", tags=["Bots"])


@router.get("")
async def get_bots():
    async with get_db() as db:
        result = await db.execute(select(Bot).where(Bot.is_deleted == False))
        return [b.to_dict() for b in result.scalars().all()]


@router.post("")
async def create_bot(bot_in: BotCreate):
    async with get_db() as db:
        bot = Bot(
            name=bot_in.name,
            account_id=bot_in.account_id,
            symbols=bot_in.symbols,
            strategy_name=bot_in.strategy_name,
            parameters=bot_in.parameters,
            status="stopped",
            # Defaults: tất cả bật khi tạo mới
            allow_new_entry=True,
            notify_entry=True,
            allow_exit_scan=True,
            notify_exit=True,
        )
        db.add(bot)
        await db.commit()
        return {"success": True, "id": bot.id}


@router.put("/{bot_id}/status")
async def update_bot_status(bot_id: int, status_update: BotStatusUpdate):
    """
    Cập nhật trạng thái bot.
    Khi chuyển sang stopped: tự động tắt allow_new_entry để không vào lệnh mới.
    Khi chuyển sang running: tự động bật lại allow_new_entry.
    """
    async with get_db() as db:
        result = await db.execute(select(Bot).where(Bot.id == bot_id, Bot.is_deleted == False))
        bot = result.scalar_one_or_none()
        if not bot:
            raise HTTPException(status_code=404, detail="Bot not found")

        new_status = status_update.status
        bot.status = new_status

        # Enforce: stopped → không vào lệnh mới
        if new_status == "stopped":
            bot.allow_new_entry = False
        elif new_status == "running":
            # Khi start lại, bật allow_new_entry (user có thể tắt lại thủ công sau)
            bot.allow_new_entry = True

        await db.commit()
        return {
            "success": True,
            "status": bot.status,
            "allow_new_entry": bot.allow_new_entry,
        }


@router.put("/{bot_id}/settings")
async def update_bot_settings(bot_id: int, settings: BotSettingsUpdate):
    """
    Cập nhật job behavior settings của bot.
    Chỉ cập nhật các field được truyền vào (partial update).
    
    - allow_new_entry: Cho phép vào lệnh mới
    - notify_entry: Gửi noti Discord khi tìm thấy entry (kể cả khi bị chặn)
    - allow_exit_scan: Quét đóng lệnh đang mở và invalidate entry opportunities
    - notify_exit: Gửi noti Discord khi đóng lệnh / invalidate entry
    """
    async with get_db() as db:
        result = await db.execute(select(Bot).where(Bot.id == bot_id, Bot.is_deleted == False))
        bot = result.scalar_one_or_none()
        if not bot:
            raise HTTPException(status_code=404, detail="Bot not found")

        if settings.allow_new_entry is not None:
            bot.allow_new_entry = settings.allow_new_entry
        if settings.notify_entry is not None:
            bot.notify_entry = settings.notify_entry
        if settings.allow_exit_scan is not None:
            bot.allow_exit_scan = settings.allow_exit_scan
        if settings.notify_exit is not None:
            bot.notify_exit = settings.notify_exit

        await db.commit()
        return {
            "success": True,
            "bot_id": bot_id,
            "allow_new_entry": bot.allow_new_entry,
            "notify_entry": bot.notify_entry,
            "allow_exit_scan": bot.allow_exit_scan,
            "notify_exit": bot.notify_exit,
        }


@router.delete("/{bot_id}")
async def delete_bot(bot_id: int):
    async with get_db() as db:
        result = await db.execute(select(Bot).where(Bot.id == bot_id))
        bot = result.scalar_one_or_none()
        if bot:
            bot.is_deleted = True
            bot.status = "stopped"
            bot.allow_new_entry = False
            await db.commit()
        return {"success": True}
